core/commands/mfs/mfs.go

```diff
@@ -0,0 +1,638 @@
+
+		if len(req.Arguments()) > 0 {
+			session := req.Arguments()[0]
+			// if a session is specified, print out just the root hash for that fs
+			root, err := node.Mfs.GetRoot(session)
+			if err != nil {
+				res.SetError(err, cmds.ErrNormal)
+				return
+			}
+
+			nd, err := root.GetValue().GetNode()
+			if err != nil {
+				res.SetError(err, cmds.ErrNormal)
+				return
+			}
+
+			k, err := nd.Key()
+			if err != nil {
+				res.SetError(err, cmds.ErrNormal)
+				return
+			}
+
+			res.SetOutput(&mfsMountListing{[]mfs.RootListing{mfs.RootListing{Hash: k}}})
+			return
+		}
+
+		// Default: list all roots
+		roots := node.Mfs.ListRoots()
+
+		res.SetOutput(&mfsMountListing{roots})
+	},
+	Marshalers: cmds.MarshalerMap{
+		cmds.Text: func(res cmds.Response) (io.Reader, error) {
+			out := res.Output().(*mfsMountListing)
+			if len(out.Mounts) == 1 && out.Mounts[0].Name == "" {
+				// a session was specified, print out just the hash
+				return strings.NewReader(out.Mounts[0].Hash.B58String()), nil
+			}
+
+			buf := new(bytes.Buffer)
+			for _, o := range out.Mounts {
+				fmt.Fprintf(buf, "%s - %s\n", o.Name, o.Hash)
+			}
+			return buf, nil
+		},
+	},
+	Type: mfsMountListing{},
+}
+
+var MfsCreateCmd = &cmds.Command{
+	Helptext: cmds.HelpText{
+		Tagline: "Create a new mutable filesystem",
+		ShortDescription: `
+Creates a new mutable filesystem based on an optional root hash.
+
+Currently, it creates a filesystem with no special publish actions,
+all changes are merely propagated to the root, and are reflected in the hash
+displayed by 'ipfs mfs'.`,
+	},
+
+	Arguments: []cmds.Argument{
+		cmds.StringArg("name", true, false, "name of filesystem to create"),
+	},
+	Options: []cmds.Option{
+		cmds.StringOption("r", "root", "root object to base new filesystem on"),
+	},
+	Run: func(req cmds.Request, res cmds.Response) {
+		node, err := req.InvocContext().GetNode()
+		if err != nil {
+			res.SetError(err, cmds.ErrNormal)
+			return
+		}
+
+		if node.Mfs == nil {
+			res.SetError(cmds.ErrNotOnline, cmds.ErrNormal)
+			return
+		}
+
+		name := req.Arguments()[0]
+		if name == "" {
+			res.SetError(errors.New("cannot have unnamed filesystem"), cmds.ErrNormal)
+			return
+		}
+
+		rtkey, found, err := req.Option("root").String()
+		if err != nil {
+			res.SetError(err, cmds.ErrNormal)
+		}
+
+		var rootnode *dag.Node
+		if found {
+			k := key.B58KeyDecode(rtkey)
+			if k == "" {
+				res.SetError(fmt.Errorf("incorrectly formatted key: %s", rtkey), cmds.ErrNormal)
+				return
+			}
+
+			nd, err := node.DAG.Get(req.Context(), k)
+			if err != nil {
+				res.SetError(err, cmds.ErrNormal)
+				return
+			}
+
+			rootnode = nd
+		} else {
+			rootnode = &dag.Node{Data: ft.FolderPBData()}
+		}
+
+		_, err = node.Mfs.NewRoot(name, rootnode, func(_ context.Context, _ key.Key) error { return nil })
+		if err != nil {
+			res.SetError(err, cmds.ErrNormal)
+			return
+		}
+	},
+}
+
+type Object struct {
+	Hash string
+}
+
+var MfsCloseCmd = &cmds.Command{
```

Comment:
i dont think this makes sense anymore. probably just `ipfs mfs rm /<rootname>`. so:
- [ ] remove `ipfs mfs close`